public/assets/ts/exam-bank.ts

```diff
@@ -0,0 +1,139 @@
+  solutionFile?: string;
+};
+
+class ExamBankFrontend {
+  static getExamList(): HTMLTableElement {
+    return document.getElementById("exams-table") as HTMLTableElement;
+  }
+
+  static init() {
+    ExamBankFrontend.populateTable();
+    (<HTMLInputElement>(
+      document.getElementById("exam-list-filter")
+    )).addEventListener("input", ExamBankFrontend.filtersInputEventListener);
+  }
+
+  static async filtersInputEventListener(this: HTMLInputElement) {
+    ExamBankFrontend.applyFiltersToTable(this.value);
+  }
+
+  /**
+   * Applies the given filters to the exam list table
+   */
+  static applyFiltersToTable(filters: string) {
+    const searchTokens = filters.split(" ");
+    const table = this.getExamList();
+
+    // parse which words filter which column
+    const courseCodeSearchTokens: string[] = [];
+    const courseDeptSearchTokens: string[] = [];
+
+    for (const token of searchTokens) {
+      if (token.match(/[0-9]/)) {
+        courseCodeSearchTokens.push(token);
+      } else {
+        courseDeptSearchTokens.push(token);
+      }
+    }
+
+    if (filters.trim().length === 0) {
+      table.classList.remove("filtered-code");
+      table.classList.remove("filtered-dept");
+    } else {
+      if (courseCodeSearchTokens.length > 0) {
+        table.classList.add("filtered-code");
+      } else {
+        table.classList.remove("filtered-code");
+      }
+      if (courseDeptSearchTokens.length > 0) {
+        table.classList.add("filtered-dept");
+      } else {
+        table.classList.remove("filtered-dept");
+      }
+    }
+
+    // apply filters
+    const examRows = Array.from(
+      table.querySelectorAll("tbody tr")
+    ) as HTMLTableRowElement[];
+
+    for (const examRow of examRows) {
+      const compareToken = (
+        tokenFromRow: string,
+        filterTokens: string[],
+        matchClass: string
+      ) => {
+        const departmentTokensMatchExam: boolean = filterTokens
+          .filter((token) => token.length > 0)
+          .reduce((foundAPreviousToken: boolean, currentToken: string) => {
+            return (
+              foundAPreviousToken ||
+              tokenFromRow.includes(currentToken.toLowerCase())
+            );
+          }, false);
+
+        if (departmentTokensMatchExam) {
+          examRow.classList.add(matchClass);
+        } else {
+          examRow.classList.remove(matchClass);
+        }
+      };
+
+      const examDepartment = examRow.getAttribute("data-course-dept");
+      compareToken(
+        examDepartment.toLowerCase(),
+        courseDeptSearchTokens,
+        "dept-match"
+      );
+
+      const examCourseCode = examRow.getAttribute("data-course-code");
+      compareToken(
+        examCourseCode.toLowerCase(),
+        courseCodeSearchTokens,
+        "code-match"
+      );
+    }
+  }
+
+  /**
+   * Builds the exam bank table
+   */
+  static async populateTable() {
+    const exams: Exam[] = await (await fetch("/api/exams")).json();
```

Comment:
uh what's this double await for

Reply:
maybe `await fetch("/api/exams").then(res => res.json())` ? double await looks funny ngl
or also 
```
const response = await fetch("/api/exams");
const exams: Exam[] = await response.json();
```

Reply:
Both `fetch` and `.json` are asynchronous.  I put the .json call into a `.then` to try to make the code more readable.